library/kani_core/src/arbitrary/pointer.rs

```diff
@@ -0,0 +1,327 @@
+        ///
+        /// # Pointer provenance
+        ///
+        /// The pointer returned in the `InBounds` and `OutOfBounds` case will have the same
+        /// provenance as the generator.
+        ///
+        /// Use the same generator if you want to handle cases where 2 or more pointers may overlap. E.g.:
+        /// ```ignore
+        /// # use kani::*;
+        /// # #[kani::proof]
+        /// # fn harness() {
+        ///     let generator = pointer_generator::<char, 5>();
+        ///     let ptr1 = generator.any_in_bounds::<char>().ptr;
+        ///     let ptr2 = generator.any_in_bounds::<char>().ptr;
+        ///     // This cover is satisfied.
+        ///     cover!(ptr1 == ptr2)
+        /// # }
+        /// ```
+        ///
+        /// If you want to cover cases where two or more pointers may not have the same
+        /// provenance, you will need to instantiate multiple generators.
+        /// You can also apply non-determinism to cover cases where the pointers may or may not
+        /// have the same provenance. E.g.:
+        ///
+        /// ```ignore
+        /// # use kani::*;
+        /// # unsafe fn my_target<T>(_ptr1: *const T; _ptr2: *const T) {}
+        /// # #[kani::proof]
+        /// # fn harness() {
+        ///     let generator1 = pointer_generator::<char, 5>();
+        ///     let generator2 = pointer_generator::<char, 5>();
+        ///     let ptr1 = generator1.any_in_bounds().ptr;
+        ///     let ptr2 = if kani::any() {
+        ///         // Pointers will have same provenance and may overlap.
+        ///         generator1.any_in_bounds().ptr;
+        ///     } else {
+        ///         // Pointers will have different provenance and will not overlap.
+        ///         generator2.any_in_bounds().ptr;
+        ///     }
+        ///     // Invoke the function under verification
+        ///     unsafe { my_target(ptr1, ptr2) };
+        /// # }
+        /// ```
+        ///
+        /// # Pointer Generator vs Pointer with any address
+        ///
+        /// Creating a pointer using the generator is different than generating a pointer
+        /// with any address.
+        ///
+        /// I.e.:
+        /// ```ignore
+        ///     // This pointer represents any address, and it may point to anything in memory,
+        ///     // allocated or not.
+        ///     let ptr1 = kani::any::<usize>() as *const u8;
+        ///
+        ///     // This pointer address will either point to unallocated memory, to a dead object
+        ///     // or to allocated memory within the generator address space.
+        ///     let generator = PointerGenerator<u8, 5>::new();
+        ///     let ptr2 = generator.any_alloc_status().ptr;
+        /// ```
+        ///
+        /// Kani cannot reason about a pointer allocation status (except for asserting its validity).
+        /// Thus, the generator was introduced to help writing harnesses that need to impose
```

Comment:
This sentence (and perhaps this entire section) needs to be moved to the beginning since it explains the whole purpose of the generator.

Reply:
I was debating about this. Writing the why before the what makes sense in proposals, but I'm not sure it reads better in the API documentation.

Reply:
What if I add one sentence to the beginning of the doc and refer to this section for more details?

Reply:
I guess it's OK. From the review perspective, it would've helped if I saw this sentence first, but from the documentation perspective, having the what before the why makes more sense.